### PR TITLE
UCM/CUDA: Added support for device-side variables.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added hook for `cudaGetSymbolSize` to support device-side variables.

### CODE REVIEW

File: src/ucm/cuda/cudamem.c [hunk 1]
```diff
@@ -66,6 +66,15 @@
         {#_func, ucm_override_##_func}, (void**)&ucm_orig_##_func \
     }
 
+#define UCM_CUDA_GET_SYMBOL_SIZE(_symbol) \
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:69]: maybe it could be a regular C function?

File: src/ucm/cuda/cudamem.c [hunk 2]
```diff
@@ -225,6 +236,11 @@ static ucm_cuda_func_t ucm_cuda_driver_funcs[] = {
 #if CUDA_VERSION >= 11020
     UCM_CUDA_FUNC_ENTRY(cuMemFreeAsync),
 #endif
+    /**
+     * cudaGetSymbolAddress is in ucm_cuda_driver_funcs list because
+     * hook for cuModuleGetGlobal doesn't intercept cudaGetSymbolAddress call.
+     * */
+    UCM_CUDA_FUNC_ENTRY(cudaGetSymbolAddress),
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:243]: why not place it in runtime_api_hooks? we install both driver and runtime hooks anyway..

File: src/ucm/cuda/cudamem.h [hunk 3]
```diff
@@ -49,5 +49,6 @@ cudaError_t ucm_cudaMallocFromPoolAsync(void **devPtr, size_t size,
                                         cudaMemPool_t memPool,
                                         cudaStream_t hStream);
 #endif
+cudaError_t ucm_cudaGetSymbolAddress(void **devPtr, const void *symbol);
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.h:52]: maybe move it before ucm_cudaMallocAsync hooks?

**Comment by yosefe** [comment]
shall we add some unit tests for this?

File: src/ucm/cuda/cudamem.c (hunk 1 above)
Comment (rakhmets) [reply src/ucm/cuda/cudamem.c]: Changed to C func.

File: src/ucm/cuda/cudamem.h (hunk 3 above)
Comment (rakhmets) [reply src/ucm/cuda/cudamem.h]: Fixed.

File: src/ucm/cuda/cudamem.c (hunk 2 above)
Comment (yosefe) [reply src/ucm/cuda/cudamem.c:234]: maybe make some changes in the hooks logic or array names then? it seems weird that a runtime function is in the "driver" API list

**Comment by rakhmets** [comment]
> shall we add some unit tests for this?

Since the feature should be tested using `__device__ ` variable, `nvcc` compiler is required. Thus, I added a test as a separate application compiled with nvcc.

File: buildlib/pr/cuda/test_malloc_hook.sh [hunk 4]
```diff
@@ -70,6 +70,9 @@ test_malloc_hook() {
 		UCX_MEM_LOG_LEVEL=diag \
 			./test/gtest/gtest --gtest_filter='cuda_hooks.*'
 
+		# Test hooks for device-side variables
+		./test/apps/test_cuda_get_symbol_address
```
Comment (brminich) [inline buildlib/pr/cuda/test_malloc_hook.sh:74]: seems test file is missing

**Comment by brminich** [comment]
CI errors are relevant

File: test/apps/test_cuda_get_symbol_address.cu [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    const int num_expected_events = 1;
+    ucp_context_h context;
+    ucs_status_t status;
+    ucp_params_t params;
+    int num_events;
+    void *dptr;
+    cudaError_t res;
+
+    params.field_mask = UCP_PARAM_FIELD_FEATURES;
+    params.features   = UCP_FEATURE_TAG | UCP_FEATURE_STREAM;
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply test/apps/test_cuda_get_symbol_address.cu]: It was just copy-pasted. Removed.

File: test/apps/test_cuda_get_symbol_address.cu [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    status            = ucp_init(&params, NULL, &context);
+    if (status != UCS_OK) {
+        printf("failed to create UCP context\n");
```
[the remark this answers is not in the document]
Comment (rakhmets) [reply test/apps/test_cuda_get_symbol_address.cu]: Fixed.

File: src/ucm/cuda/cudamem.c [hunk 7]
```diff
@@ -228,6 +257,17 @@ static ucm_cuda_func_t ucm_cuda_driver_funcs[] = {
     {{NULL}, NULL}
 };
 
+static size_t ucm_cuda_get_symbol_size(const void *symbol)
+{
+    size_t size;
+
+    if (cudaGetSymbolSize(&size, symbol) != cudaSuccess) {
```
Comment (brminich) [inline src/ucm/cuda/cudamem.c:264]: should it be?
```suggestion
    if (ucm_orig_cudaGetSymbolSize(&size, symbol) != cudaSuccess) {
```

Comment (rakhmets) [reply src/ucm/cuda/cudamem.c:264]: No, we do not change `cudaGetSymbolSize`. Thus, `ucm_orig_cudaGetSymbolSize` is not defined.

**Review by brminich — APPROVED** [review]

**Review by brminich — APPROVED** [review]

File: config/m4/cuda.m4 [hunk 8]
```diff
@@ -78,6 +81,8 @@ AS_IF([test "x$cuda_checked" != "xyes"],
                         [AC_DEFINE([HAVE_CUDA_FABRIC], 1, [Enable CUDA fabric handle support])],
                         [], [[#include <cuda.h>]])
 
+         AC_PATH_PROGS(NVCC, nvcc, "", $PATH:$CUDA_BIN_PATH)
```
Comment (yosefe) [inline config/m4/cuda.m4:84]: maybe
$CUDA_BIN_PATH:$PATH
to prefer the with arg over env var?

File: src/ucm/cuda/cudamem.c [hunk 9]
```diff
@@ -87,6 +87,8 @@ UCM_DEFINE_REPLACE_DLSYM_PTR_FUNC(cuMemAllocPitch_v2, CUresult, -1,
 UCM_DEFINE_REPLACE_DLSYM_PTR_FUNC(cuMemMap, CUresult, -1, CUdeviceptr, size_t,
                                   size_t, CUmemGenericAllocationHandle,
                                   unsigned long long)
+UCM_DEFINE_REPLACE_DLSYM_PTR_FUNC(cuModuleGetGlobal_v2, CUresult, -1,
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:90]: maybe patch also cuModuleGetGlobal (the v1)?

File: src/ucm/cuda/cudamem.c [hunk 10]
```diff
@@ -206,13 +210,38 @@ UCM_CUDA_FREE_FUNC(cuMemFreeAsync, UCS_MEMORY_TYPE_CUDA, CUresult, arg0, 0,
                    "ptr=0x%llx, stream=%p", CUdeviceptr, CUstream)
 #endif
 
+CUresult ucm_cuModuleGetGlobal_v2(CUdeviceptr *dptr, size_t *bytes,
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:213]: why couldn't reuse UCM_CUDA_ALLOC_FUNC?

File: config/m4/cuda.m4 (hunk 8 above)
Comment (rakhmets) [reply config/m4/cuda.m4]: Fixed.

File: src/ucm/cuda/cudamem.c [hunk 11]
```diff
@@ -206,13 +210,38 @@ UCM_CUDA_FREE_FUNC(cuMemFreeAsync, UCS_MEMORY_TYPE_CUDA, CUresult, arg0, 0,
...
+
+    if (dptr == NULL) {
+        return ucm_orig_cuModuleGetGlobal_v2(dptr, bytes, hmod, name);
+    }
+
+    size_ptr = (bytes == NULL) ? &size : bytes;
+    ucm_event_enter();
+    ret = ucm_orig_cuModuleGetGlobal_v2(dptr, size_ptr, hmod, name);
+    if (ret == CUDA_SUCCESS) {
+        ucm_trace("%s(bytes=%p, hmod=%p, name=%s) returned %p", __func__, bytes,
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:228]: "returned %p" -> "returned dptr=%p"
bytes -> size_ptr

File: src/ucm/cuda/cudamem.c [hunk 12]
```diff
@@ -206,13 +210,38 @@ UCM_CUDA_FREE_FUNC(cuMemFreeAsync, UCS_MEMORY_TYPE_CUDA, CUresult, arg0, 0,
...
+                  hmod, name, (void*)(*dptr));
+        ucm_cuda_dispatch_mem_alloc(*dptr, *size_ptr);
+    }
+
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:232]: minot: remove space line

File: src/ucm/cuda/cudamem.c [hunk 13]
```diff
@@ -206,13 +210,38 @@ UCM_CUDA_FREE_FUNC(cuMemFreeAsync, UCS_MEMORY_TYPE_CUDA, CUresult, arg0, 0,
...
 
+CUresult ucm_cuModuleGetGlobal_v2(CUdeviceptr *dptr, size_t *bytes,
+                                  CUmodule hmod, const char *name)
+{
+    CUresult ret;
+    size_t size;
+    size_t *size_ptr;
+
+    if (dptr == NULL) {
+        return ucm_orig_cuModuleGetGlobal_v2(dptr, bytes, hmod, name);
```
Comment (yosefe) [inline src/ucm/cuda/cudamem.c:221]: add comment:
/* Device pointer is not returned */

Comment (rakhmets) [reply src/ucm/cuda/cudamem.c:222]: Added.

File: src/ucm/cuda/cudamem.c (hunk 12 above)
Comment (rakhmets) [reply src/ucm/cuda/cudamem.c:233]: But we have the recommendation in `The UCX code style` to use blank line after if/else blocks.

File: src/ucm/cuda/cudamem.c (hunk 11 above)
Comment (rakhmets) [reply src/ucm/cuda/cudamem.c]: Updated.

File: src/ucm/cuda/cudamem.c (hunk 12 above)
Comment (yosefe) [reply src/ucm/cuda/cudamem.c:233]: it seems inconsistent with ucm_event_enter

Comment (rakhmets) [reply src/ucm/cuda/cudamem.c]: Removed.

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Pull request merged**